### DISCUSSION TITLE
Using "simulated_audio_pot" with encoder

### DISCUSSION THREAD

**Discussion by nradina** [pschatzmann/arduino-audio-tools#568] [Q&A] [answered]
Hello everyone,

I'm trying to wrap my head around using the excellent volume controls in this amazing library.

I've read through the[ "SimulatedAudioPot Class Reference" ](https://pschatzmann.github.io/arduino-audio-tools/html/classaudio__tools_1_1_simulated_audio_pot.html#a4adec6dfd3db79b4d459a6616d376dc2)yet I'm unsure how to implement this with a digital rotary encoder.

Are there any code examples that would put me on a better track to implement a rotary encoder to control volume of the AudioKitStream output?

I started with this code yet I'm sure this is not correct nor does it reference the <[VolumeControl.h](https://pschatzmann.github.io/arduino-audio-tools/html/_volume_control_8h_source.html)> library.

`/**
   @file basic-a2dp-audiokit.ino
   @brief see https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/basic-a2dp-audiokit/README.md

   @author Phil Schatzmann
   @copyright GPLv3
*/

#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"
#include "AudioLibs/AudioA2DP.h"
#include <Rotary.h>

// Pin definitions for the rotary encoder
#define ROTARY_ENCODER_A 21
#define ROTARY_ENCODER_B 22

// The rotary encoder object
Rotary rotaryEncoder = Rotary(ROTARY_ENCODER_A, ROTARY_ENCODER_B);

BluetoothA2DPSink a2dp_sink;
AudioKitStream kit;

// Write data to AudioKit in callback
void read_data_stream(const uint8_t *data, uint32_t length) {
  kit.write(data, length);
}

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // setup output
  auto cfg = kit.defaultConfig(TX_MODE);
  cfg.sd_active = false;
  kit.begin(cfg);

  // register callback
  a2dp_sink.set_stream_reader(read_data_stream, false);
  a2dp_sink.start("AudioKit");
}

void loop() {
  kit.processActions();

  // Get the number of steps taken by the encoder
  int steps = rotaryEncoder.process();

  // check if steps !=0
  if (steps) {
    // Check the direction of rotation and call the appropriate function
    if (steps > 0) {
      // Rotate clockwise
      for (int i = 0; i < steps; i++) {
        kit.incrementVolume();
      }
    } else {
      // Rotate counterclockwise
      for (int i = 0; i < -steps; i++) {
        kit.decrementVolume();
      }
    }
  }
  delay(10);
}

'

**Comment by pschatzmann** [comment, marked as answer]
kit.setVolume(), incrementVolume(), decrementVolume() is using the built-in volume functionality of the Audiokit.
You should not add any delay in the loop because this might disturb the audio output.
Why didn't you just use  kit.incrementVolume(steps);  I would expect that this should work correctly both for positive and negative values...

**Reply by nradina** [reply]
Thank you, Phil. In order to use "kit.incrementVolume(steps);" with an encoder, do I need to include an encoder library or is this not necessary with your library?

**Reply by pschatzmann** [reply]
I never used any Rotary Encoder myself and have no functionality for this in my library.
So yes you need to use an external library...
